{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Files in data/ must be downloaded from https://doi.org/10.5281/zenodo.10932490\n",
    "graph_A:str = 'data/15_yeast_mgc_CASBJU01_chr_1.gfa'\n",
    "graph_B:str = 'data/15_yeast_pggb_chr_1.gfa'\n",
    "# Files in temp/* will be created during the execution of the notebook\n",
    "edition_file: str = 'temp/yeast_edit.json'\n",
    "variants_file_A:str = \"temp/cactus.vcf\" # path to a vcf for graph A\n",
    "variants_file_B:str = \"temp/pggb.vcf\" # path to a vcf for graph B\n",
    "variants_filter_A:str = \"temp/cactus.filter.vcf\" # path to a vcf for graph A\n",
    "variants_filter_B:str = \"temp/pggb.filter.vcf\" # path to a vcf for graph B\n",
    "private_variants_file_A:str = 'temp/cactus.private.vcf'\n",
    "private_variants_file_B:str = 'temp/pggb.private.vcf'\n",
    "reference:list[str] = ['CASBJU01'] # choosing a reference to plot against\n",
    "chromosom:str = '1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colorblind-friendly palette\n",
    "colors:dict[str,str] = {\n",
    "    'blue':    '#377eb8', \n",
    "    'orange':  '#ff7f00',\n",
    "    'green':   '#4daf4a',\n",
    "    'pink':    '#f781bf',\n",
    "    'brown':   '#a65628',\n",
    "    'purple':  '#984ea3',\n",
    "    'gray':    '#999999',\n",
    "    'red':     '#e41a1c',\n",
    "    'yellow':  '#dede00'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from json import load\n",
    "from matplotlib import pyplot as plt\n",
    "from numpy import nan\n",
    "from pgGraphs import Graph\n",
    "from os import system\n",
    "from seaborn import regplot\n",
    "from pandas import DataFrame\n",
    "from scipy import stats\n",
    "from itertools import pairwise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating mandatory files to compare graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, we compute edition to be sure to have a list of breakpoints\n",
    "system(f'pancat edit {graph_A} {graph_B} -o {edition_file}')\n",
    "\n",
    "# Then, we compute variants file\n",
    "system(f'vg deconstruct -e -a -p \"{reference[0]}#{reference[0]}\" {graph_A} > {variants_file_A}') # format for raw cactus files > \"seq000#{reference[0]}\"\n",
    "system(f'vg deconstruct -e -a -p \"{reference[0]}\" {graph_B} > {variants_file_B}') # format for pggb files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading files in memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "editions:dict[str,list] = load(open(edition_file,'r',encoding='utf-8'))\n",
    "\n",
    "cactus_gfa:Graph = Graph(graph_A)\n",
    "pggb_gfa:Graph = Graph(graph_B)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Showing the different path names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cactus_gfa.paths.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performing comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we will look at the distribution of variants along the graph.\n",
    "- We need to get the length of the reference sequence for the x-axis of the graph\n",
    "- Then, we can plot by position the variants from the according vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can take any of the two as they have the same content in sequences\n",
    "reference_length:int = sum([cactus_gfa.segments[x]['length'] for x,_ in cactus_gfa.paths[reference[0].upper()]['path']])\n",
    "print(f'Length of the reference is {reference_length} bp.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We open the variants files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_A_variants:dict[str,tuple] = dict()\n",
    "with open(variants_file_A,'r') as vcf_A:\n",
    "    for line in vcf_A:\n",
    "        if not line.startswith('#'):\n",
    "            l:list[str] = line.split()\n",
    "            file_A_variants[l[1]]=(l[3],l[4])\n",
    "            \n",
    "file_B_variants:dict[str,tuple] = dict()\n",
    "with open(variants_file_B,'r') as vcf_B:\n",
    "    for line in vcf_B:\n",
    "        if not line.startswith('#'):\n",
    "            l:list[str] = line.split()\n",
    "            file_B_variants[l[1]]=(l[3],l[4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We split the variants in batches (percentage of the sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling:int = 200\n",
    "\n",
    "# Comptage des variants dans chacun des graphes (variants totaux)\n",
    "counts_A:list[int] = [0 for _ in range(sampling)]\n",
    "counts_B:list[int] = [0 for _ in range(sampling)]\n",
    "counts_A_sup_2:list[int] = [0 for _ in range(sampling)]\n",
    "counts_B_sup_2:list[int] = [0 for _ in range(sampling)]\n",
    "counts_A_sup_50:list[int] = [0 for _ in range(sampling)]\n",
    "counts_B_sup_50:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "# Comptage des variants privÃ©s\n",
    "p_counts_A:list[int] = [0 for _ in range(sampling)]\n",
    "p_counts_B:list[int] = [0 for _ in range(sampling)]\n",
    "p_counts_A_sup_2:list[int] = [0 for _ in range(sampling)]\n",
    "p_counts_B_sup_2:list[int] = [0 for _ in range(sampling)]\n",
    "p_counts_A_sup_50:list[int] = [0 for _ in range(sampling)]\n",
    "p_counts_B_sup_50:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "\n",
    "ratios:float = sampling/reference_length\n",
    "\n",
    "\n",
    "for pos in file_A_variants.keys():\n",
    "    is_common_variant:bool = file_B_variants.get(pos,None) == file_A_variants.get(pos,None)\n",
    "    counts_A[int(int(pos)*ratios)] += 1\n",
    "    if not is_common_variant:\n",
    "        p_counts_A[int(int(pos)*ratios)] += 1\n",
    "    if any([len(x) >= 50 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "        counts_A_sup_50[int(int(pos)*ratios)] += 1\n",
    "        if not is_common_variant:\n",
    "            p_counts_A_sup_50[int(int(pos)*ratios)] += 1\n",
    "    if any([len(x) >= 2 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "        counts_A_sup_2[int(int(pos)*ratios)] += 1\n",
    "        if not is_common_variant:\n",
    "            p_counts_A_sup_2[int(int(pos)*ratios)] += 1\n",
    "\n",
    "for pos in file_B_variants.keys():\n",
    "    is_common_variant:bool = file_B_variants.get(pos,None) == file_A_variants.get(pos,None)\n",
    "    counts_B[int(int(pos)*ratios)] += 1\n",
    "    if not is_common_variant:\n",
    "        p_counts_B[int(int(pos)*ratios)] += 1\n",
    "    if any([len(x) >= 50 for x in [file_B_variants[pos][0]] + file_B_variants[pos][1].split(',')]):\n",
    "        counts_B_sup_50[int(int(pos)*ratios)] += 1\n",
    "        if not is_common_variant:\n",
    "            p_counts_B_sup_50[int(int(pos)*ratios)] += 1\n",
    "    if any([len(x) >= 2 for x in [file_B_variants[pos][0]] + file_B_variants[pos][1].split(',')]):\n",
    "        counts_B_sup_2[int(int(pos)*ratios)] += 1\n",
    "        if not is_common_variant:\n",
    "            p_counts_B_sup_2[int(int(pos)*ratios)] += 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We plot variations along the reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we compute editions along the reference. We can only display editions on reference, sadly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling:int = 200\n",
    "counts_editions:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "counts_merges:list[int] = [0 for _ in range(sampling)]\n",
    "counts_splits:list[int] = [0 for _ in range(sampling)]\n",
    "ratios:float = sampling/reference_length\n",
    "\n",
    "for merge in editions[reference[0].upper()]['merges']:\n",
    "    counts_editions[int(float(merge[0])*ratios)] += 1\n",
    "\n",
    "for split in editions[reference[0].upper()]['splits']:\n",
    "    counts_editions[int(int(split[0])*ratios)] += 1\n",
    "    \n",
    "for merge in editions[reference[0].upper()]['merges']:\n",
    "    counts_merges[int(float(merge[0])*ratios)] += 1\n",
    "\n",
    "for split in editions[reference[0].upper()]['splits']:\n",
    "    counts_splits[int(int(split[0])*ratios)] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_nodes_cactus:list[int] = [0 for _ in range(sampling)]\n",
    "counts_nodes_pggb:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "pos_counter: int = 0\n",
    "for x,_ in cactus_gfa.paths[reference[0].upper()]['path']:\n",
    "    counts_nodes_cactus[int(float(pos_counter)*ratios)] +=1\n",
    "    pos_counter += cactus_gfa.segments[x]['length']\n",
    "    \n",
    "pos_counter: int = 0\n",
    "for x,_ in pggb_gfa.paths[reference[0].upper()]['path']:\n",
    "    counts_nodes_pggb[int(float(pos_counter)*ratios)] +=1\n",
    "    pos_counter += pggb_gfa.segments[x]['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import rc\n",
    "\n",
    "font = {'size': 16}\n",
    "\n",
    "rc('font', **font)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(18,5))\n",
    "#fig.suptitle(f\"Distribution of variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "#fig.supylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "ax.fill_between(list(range(sampling)),counts_editions, color= \"darkred\", label='merges')\n",
    "ax.fill_between(list(range(sampling)),counts_splits, color= \"gold\", label='splits')\n",
    "ax.set_ylabel(f'Editions per {round((reference_length/200)/1000)}kb.')\n",
    "#ax.set_xticklabels([round(__/ratios) if not __%20 else '' for __ in range(sampling)])\n",
    "ax.legend(loc='upper left')\n",
    "\n",
    "\n",
    "\n",
    "axb = ax.twinx()\n",
    "axb.set_ylabel(f'Nodes per {round((reference_length/200)/1000)}kb.')\n",
    "axb.set_xticks([__ for __ in range(sampling+1) if not __%20])\n",
    "axb.set_xticklabels([int(round(__/ratios,-3)) for __ in range(sampling+1) if not __%20 ])\n",
    "\n",
    "\n",
    "axb.plot(counts_nodes_cactus,color='seagreen',linestyle='dashed', label='MGC')\n",
    "axb.plot(counts_nodes_pggb,color='indianred',linestyle='dashed', label='PGGB')\n",
    "axb.legend(loc='upper right')\n",
    "\n",
    "plt.savefig(\"figures/editions_nodes_along_reference.png\", bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4), (ax5, ax6)) = plt.subplots(3,2, sharex=True, sharey='row', figsize=(24,10))\n",
    "fig.suptitle(f\"Distribution of variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "#fig.supylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "alpha = 0.5\n",
    "\n",
    "for axs in (ax1, ax2, ax3, ax4, ax5, ax6):\n",
    "    __ = axs.twinx()\n",
    "    __.fill_between(list(range(sampling)),counts_editions, color= \"violet\",alpha= alpha)\n",
    "    __.set_ylabel(f'Editions per {round((reference_length/200)/1000)}kb.')\n",
    "    axs.set_xticks([])\n",
    "    axs.set_ylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "ax1.set_title(f\"Total variants for mgc graph (total: {sum(counts_A)})\")\n",
    "ax1.plot(counts_A)\n",
    "ax1.plot(counts_A_sup_2)\n",
    "ax1.plot(counts_A_sup_50)\n",
    "\n",
    "\n",
    "ax2.set_title(f\"Total variants for pggb graph (total: {sum(counts_B)})\")\n",
    "ax2.plot(counts_B, label='$\\geq$1bp')\n",
    "ax2.plot(counts_B_sup_2, label='$\\geq$2bp')\n",
    "ax2.plot(counts_B_sup_50, label='$\\geq$50bp')\n",
    "\n",
    "ax2.legend(loc='upper right')\n",
    "\n",
    "ax3.set_title(f\"Private variants for mgc graph (total: {sum(p_counts_A)})\")\n",
    "ax3.plot(p_counts_A)\n",
    "ax3.plot(p_counts_A_sup_2)\n",
    "ax3.plot(p_counts_A_sup_50)\n",
    "\n",
    "ax4.set_title(f\"Private variants for pggb graph (total: {sum(p_counts_B)})\")\n",
    "ax4.plot(p_counts_B, label='$\\geq$1bp')\n",
    "ax4.plot(p_counts_B_sup_2, label='$\\geq$2bp')\n",
    "ax4.plot(p_counts_B_sup_50, label='$\\geq$50bp')\n",
    "\n",
    "ax4.legend(loc='upper right')\n",
    "\n",
    "ax5.set_title(f\"Shared variants (total: {sum(counts_A)-sum(p_counts_A)})\")\n",
    "ax5.plot([counts_A[i] - p_counts_A[i] for i in range(sampling)])\n",
    "ax5.plot([counts_A_sup_2[i] - p_counts_A_sup_2[i] for i in range(sampling)])\n",
    "ax5.plot([counts_A_sup_50[i] - p_counts_A_sup_50[i] for i in range(sampling)])\n",
    "\n",
    "ax6.set_title(f\"Summation of private variants (total: {sum(p_counts_A)+sum(p_counts_B)})\")\n",
    "ax6.plot([p_counts_A[i]+ p_counts_B[i] for i in range(sampling)], label='$\\geq$1bp')\n",
    "ax6.plot([p_counts_A_sup_2[i]+ p_counts_B_sup_2[i] for i in range(sampling)], label='$\\geq$2bp')\n",
    "ax6.plot([p_counts_A_sup_50[i] +p_counts_B_sup_50[i] for i in range(sampling)], label='$\\geq$50bp')\n",
    "\n",
    "ax6.legend(loc='upper right')\n",
    "\n",
    "plt.savefig(\"figures/baseline_profile.png\", bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, (ax2, ax5, ax6) = plt.subplots(3,1, sharex=True, sharey='row', figsize=(17,13))\n",
    "\n",
    "#fig.supylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "alpha = 0.3\n",
    "\n",
    "for axs in (ax2, ax5, ax6):\n",
    "    __ = axs.twinx()\n",
    "    __.fill_between(list(range(sampling)),counts_editions, color= \"violet\",alpha=alpha)\n",
    "    __.set_ylabel(f'Editions per {round((reference_length/200)/1000)}kb.')\n",
    "    axs.set_xticks([])\n",
    "    axs.set_ylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "\n",
    "ax2.set_title(f\"Total variants across both graphs (MGC: {sum(counts_A)}, PGGB: {sum(counts_B)})\")\n",
    "ax2.plot(counts_A, label='MGC', color='seagreen', linewidth=3)\n",
    "ax2.plot(counts_B, label='PGGB', color='indianred', linewidth=3)\n",
    "\n",
    "ax2.legend(loc='upper right')\n",
    "\n",
    "ax5.set_title(f\"Shared variants across both graphs (total: {sum(counts_A)-sum(p_counts_A)})\")\n",
    "ax5.plot([counts_A[i] - p_counts_A[i] for i in range(sampling)], label='$\\geq$1bp', color='dodgerblue', linewidth=3)\n",
    "#ax5.plot([counts_A_sup_2[i] - p_counts_A_sup_2[i] for i in range(sampling)], label='$\\geq$2bp')\n",
    "#ax5.plot([counts_A_sup_50[i] - p_counts_A_sup_50[i] for i in range(sampling)], label='$\\geq$50bp')\n",
    "\n",
    "#ax5.legend(loc='upper right')\n",
    "\n",
    "ax6.set_title(f\"Summation of private variants (total: {sum(p_counts_A)+sum(p_counts_B)})\")\n",
    "ax6.plot([p_counts_A[i]+ p_counts_B[i] for i in range(sampling)], label='$\\geq$1bp', color='orangered', linewidth=3)\n",
    "#ax6.plot([p_counts_A_sup_2[i]+ p_counts_B_sup_2[i] for i in range(sampling)], label='$\\geq$2bp')\n",
    "#ax6.plot([p_counts_A_sup_50[i] +p_counts_B_sup_50[i] for i in range(sampling)], label='$\\geq$50bp')\n",
    "ax6.set_xticks([__ for __ in range(sampling+1) if not __%20])\n",
    "ax6.set_xticklabels([int(round(__/ratios,-3)) for __ in range(sampling+1) if not __%20 ])\n",
    "\n",
    "#ax6.legend(loc='upper right')\n",
    "\n",
    "plt.savefig(\"figures/baseline_profile_supp.png\", bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4), (ax5, ax6)) = plt.subplots(3,2, sharex=True, sharey='row', figsize=(16,10))\n",
    "fig.suptitle(f\"Distribution of variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "#fig.supylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "alpha = 0.5\n",
    "\n",
    "for axs in (ax1, ax2, ax3, ax4, ax5, ax6):\n",
    "    __ = axs.twinx()\n",
    "    __.fill_between(list(range(sampling)),counts_editions, color= \"violet\",alpha= alpha)\n",
    "    __.set_ylabel(f'Editions per {round((reference_length/200)/1000)}kb.')\n",
    "    axs.set_xticks([])\n",
    "    axs.set_ylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "ax1.set_title(f\"Total variants for mgc graph (total: {sum(counts_A)})\")\n",
    "ax1.plot(counts_A_sup_50)\n",
    "\n",
    "\n",
    "ax2.set_title(f\"Total variants for pggb graph (total: {sum(counts_B)})\")\n",
    "ax2.plot(counts_B_sup_50, label='$\\geq$50bp')\n",
    "\n",
    "ax2.legend(loc='upper right')\n",
    "\n",
    "ax3.set_title(f\"Private variants for mgc graph (total: {sum(p_counts_A)})\")\n",
    "ax3.plot(p_counts_A_sup_50)\n",
    "\n",
    "ax4.set_title(f\"Private variants for pggb graph (total: {sum(p_counts_B)})\")\n",
    "ax4.plot(p_counts_B_sup_50, label='$\\geq$50bp')\n",
    "\n",
    "ax4.legend(loc='upper right')\n",
    "\n",
    "ax5.set_title(f\"Shared variants (total: {sum(counts_A)-sum(p_counts_A)})\")\n",
    "ax5.plot([counts_A_sup_50[i] - p_counts_A_sup_50[i] for i in range(sampling)])\n",
    "\n",
    "ax6.set_title(f\"Summation of private variants (total: {sum(p_counts_A)+sum(p_counts_B)})\")\n",
    "ax6.plot([p_counts_A_sup_50[i] +p_counts_B_sup_50[i] for i in range(sampling)], label='$\\geq$50bp')\n",
    "\n",
    "ax6.legend(loc='upper right')\n",
    "\n",
    "plt.savefig(\"figures/baseline_profile.png\", bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We build a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ve:DataFrame = DataFrame(\n",
    "    data={\n",
    "        'edition_count':counts_editions,\n",
    "        'full_total':[(counts_A[i]+counts_B[i])/2 for i in range(sampling)],\n",
    "        'total_variants_mgc':counts_A,\n",
    "        'total_variants_mgc_sup_2':counts_A_sup_2,\n",
    "        'total_variants_mgc_sup_50':counts_A_sup_50,\n",
    "        'private_variants_mgc':p_counts_A,\n",
    "        'private_variants_mgc_sup_2':p_counts_A_sup_2,\n",
    "        'private_variants_mgc_sup_50':p_counts_A_sup_50,\n",
    "        'total_variants_pggb':counts_B,\n",
    "        'total_variants_pggb_sup_2':counts_B_sup_2,\n",
    "        'total_variants_pggb_sup_50':counts_B_sup_50,\n",
    "        'private_variants_pggb':p_counts_B,\n",
    "        'private_variants_pggb_sup_2':p_counts_B_sup_2,\n",
    "        'private_variants_pggb_sup_50':p_counts_B_sup_50,\n",
    "        'shared_variants':[counts_A[i] - p_counts_A[i] for i in range(sampling)],\n",
    "        'shared_variants_sup_2':[counts_A_sup_2[i] - p_counts_A_sup_2[i] for i in range(sampling)],\n",
    "        'shared_variants_sup_50':[counts_A_sup_50[i] - p_counts_A_sup_50[i] for i in range(sampling)],\n",
    "        'private_variants_sum':[p_counts_A[i]+ p_counts_B[i] for i in range(sampling)],\n",
    "        'private_variants_sum_sup_2':[p_counts_A_sup_2[i]+ p_counts_B_sup_2[i] for i in range(sampling)],\n",
    "        'private_variants_sum_sup_50':[p_counts_A_sup_50[i] +p_counts_B_sup_50[i] for i in range(sampling)],\n",
    "        'mgc_nodes':counts_nodes_cactus,\n",
    "        'pggb_nodes':counts_nodes_pggb,\n",
    "        'average_nodes':[(counts_nodes_cactus[i]+counts_nodes_pggb[i])/2 for i in range(sampling)]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we build correlograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5,5))\n",
    "\n",
    "\n",
    "ax.set_title(f\"Total: r={round(stats.pearsonr(df_ve['average_nodes'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['average_nodes'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='mgc_nodes', y='edition_count', data=df_ve, ax=ax, color='seagreen')\n",
    "regplot(x='pggb_nodes', y='edition_count', data=df_ve, ax=ax, color='indianred')\n",
    "ax.set_ylabel('Editions')\n",
    "ax.set_xlabel('Nodes')\n",
    "\n",
    "plt.savefig(\"figures/correlograms_nodes.png\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, (ax2, ax3, ax4) = plt.subplots(3,1, figsize=(4,13))\n",
    "#fig.suptitle(f\"Correlation between editions and variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "ax2.set_title(f\"Total: r={round(stats.pearsonr(df_ve['full_total'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['full_total'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='total_variants_mgc', y='edition_count', data=df_ve, ax=ax2, color='seagreen')\n",
    "regplot(x='total_variants_pggb', y='edition_count', data=df_ve, ax=ax2, color='indianred')\n",
    "ax2.set_ylabel('Editions')\n",
    "ax2.set_xlabel('')\n",
    "\n",
    "ax3.set_title(f\"Shared: r={round(stats.pearsonr(df_ve['shared_variants'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['shared_variants'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='shared_variants', y='edition_count', data=df_ve, ax=ax3, color='dodgerblue')\n",
    "#regplot(x='shared_variants_sup_2', y='edition_count', data=df_ve, ax=ax3)\n",
    "#regplot(x='shared_variants_sup_50', y='edition_count', data=df_ve, ax=ax3)\n",
    "ax3.set_ylabel('Editions')\n",
    "ax3.set_xlabel('')\n",
    "\n",
    "ax4.set_title(f\"Private: r={round(stats.pearsonr(df_ve['private_variants_sum'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['private_variants_sum'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='private_variants_sum', y='edition_count', data=df_ve, ax=ax4, color='orangered')\n",
    "#regplot(x='private_variants_sum_sup_2', y='edition_count', data=df_ve, ax=ax4)\n",
    "#regplot(x='private_variants_sum_sup_50', y='edition_count', data=df_ve, ax=ax4)\n",
    "ax4.set_ylabel('Editions')\n",
    "ax4.set_xlabel('')\n",
    "\n",
    "plt.savefig(\"figures/correlograms_editions.png\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2, figsize=(14,14))\n",
    "fig.suptitle(f\"Correlation between editions and variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "\n",
    "ax1.set_title(f\"Private to mgc: r={round(stats.pearsonr(df_ve['private_variants_mgc'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['private_variants_mgc'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='private_variants_mgc', y='edition_count', data=df_ve, ax=ax1)\n",
    "regplot(x='private_variants_mgc_sup_2', y='edition_count', data=df_ve, ax=ax1)\n",
    "regplot(x='private_variants_mgc_sup_50', y='edition_count', data=df_ve, ax=ax1)\n",
    "\n",
    "ax2.set_title(f\"Private to pggb: r={round(stats.pearsonr(df_ve['private_variants_pggb'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['private_variants_pggb'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='private_variants_pggb', y='edition_count', data=df_ve, ax=ax2)\n",
    "regplot(x='private_variants_pggb_sup_2', y='edition_count', data=df_ve, ax=ax2)\n",
    "regplot(x='private_variants_pggb_sup_50', y='edition_count', data=df_ve, ax=ax2)\n",
    "\n",
    "ax3.set_title(f\"Shared: r={round(stats.pearsonr(df_ve['shared_variants'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['shared_variants'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='shared_variants', y='edition_count', data=df_ve, ax=ax3)\n",
    "regplot(x='shared_variants_sup_2', y='edition_count', data=df_ve, ax=ax3)\n",
    "regplot(x='shared_variants_sup_50', y='edition_count', data=df_ve, ax=ax3)\n",
    "\n",
    "ax4.set_title(f\"Private: r={round(stats.pearsonr(df_ve['private_variants_sum'], df_ve['edition_count'])[0],2)}, p={stats.pearsonr(df_ve['private_variants_sum'], df_ve['edition_count'])[1]:.2e}\")\n",
    "regplot(x='private_variants_sum', y='edition_count', data=df_ve, ax=ax4)\n",
    "regplot(x='private_variants_sum_sup_2', y='edition_count', data=df_ve, ax=ax4)\n",
    "regplot(x='private_variants_sum_sup_50', y='edition_count', data=df_ve, ax=ax4)\n",
    "\n",
    "plt.savefig(\"figures/correlograms_variants_classes.png\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling:int = 200\n",
    "\n",
    "# Comptage des variants dans chacun des graphes (variants totaux)\n",
    "counts_A_substitutions:list[int] = [0 for _ in range(sampling)]\n",
    "counts_B_substitutions:list[int] = [0 for _ in range(sampling)]\n",
    "counts_A_private_substitution:list[int] = [0 for _ in range(sampling)]\n",
    "counts_B_private_substitution:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "shared_others:list[int] = [0 for _ in range(sampling)]\n",
    "shared_size_1:list[int] = [0 for _ in range(sampling)]\n",
    "shared_size_2_9:list[int] = [0 for _ in range(sampling)]\n",
    "shared_size_10_49:list[int] = [0 for _ in range(sampling)]\n",
    "shared_size_50:list[int] = [0 for _ in range(sampling)]\n",
    "\n",
    "ratios:float = sampling/reference_length\n",
    "\n",
    "with open(private_variants_file_A,'w',encoding='utf-8') as pwriter_A:\n",
    "    with open(variants_file_A,'r',encoding='utf-8') as reader_A:\n",
    "        all_lines:list[str] = reader_A.readlines()\n",
    "        for line in all_lines:\n",
    "            if line.startswith('#'):\n",
    "                pwriter_A.write(line)\n",
    "        for pos in file_A_variants.keys():\n",
    "            is_common_variant:bool = file_B_variants.get(pos,None) == file_A_variants.get(pos,None)\n",
    "            if any([len(x) == len(file_A_variants[pos][0]) for x in file_A_variants[pos][1].split(',')]):\n",
    "                counts_A_substitutions[int(int(pos)*ratios)] += 1\n",
    "                if not is_common_variant:\n",
    "                    counts_A_private_substitution[int(int(pos)*ratios)] += 1\n",
    "            if is_common_variant:\n",
    "                if all([len(x) == 1 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "                    shared_size_1[int(int(pos)*ratios)] += 1\n",
    "                elif all([len(x) >= 2 and len(x) < 10 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "                    shared_size_2_9[int(int(pos)*ratios)] += 1\n",
    "                elif all([len(x) >= 10 and len(x) < 50 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "                    shared_size_10_49[int(int(pos)*ratios)] += 1\n",
    "                elif all([len(x) >= 50 for x in [file_A_variants[pos][0]] + file_A_variants[pos][1].split(',')]):\n",
    "                    shared_size_50[int(int(pos)*ratios)] += 1\n",
    "                else:\n",
    "                    # We have a variant with multiple sizes\n",
    "                    shared_others[int(int(pos)*ratios)] += 1\n",
    "            else:\n",
    "                for line in all_lines:\n",
    "                    if not line.startswith('#'):\n",
    "                        l:list[str] = line.split()\n",
    "                        if pos == l[1] and file_A_variants[pos] == (l[3],l[4]):\n",
    "                            pwriter_A.write(line)\n",
    "                    \n",
    "with open(private_variants_file_B,'w',encoding='utf-8') as pwriter_B:\n",
    "    with open(variants_file_B,'r',encoding='utf-8') as reader_B:\n",
    "        all_lines:list[str] = reader_B.readlines()\n",
    "        for line in all_lines:\n",
    "            if line.startswith('#'):\n",
    "                pwriter_B.write(line)\n",
    "        for pos in file_B_variants.keys():\n",
    "            is_common_variant:bool = file_B_variants.get(pos,None) == file_A_variants.get(pos,None)\n",
    "            if any([len(x) == len(file_B_variants[pos][0]) for x in file_B_variants[pos][1].split(',')]):\n",
    "                counts_B_substitutions[int(int(pos)*ratios)] += 1\n",
    "                if not is_common_variant:\n",
    "                    counts_B_private_substitution[int(int(pos)*ratios)] += 1\n",
    "                    for line in all_lines:\n",
    "                        if not line.startswith('#'):\n",
    "                            l:list[str] = line.split()\n",
    "                            if pos == l[1] and file_B_variants[pos] == (l[3],l[4]):\n",
    "                                pwriter_B.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2),(ax3,ax4)) = plt.subplots(2,2, sharex=True, sharey='row', figsize=(16,7))\n",
    "fig.suptitle(f\"Distribution of variants between graphs from MGC and PGGB for chromosom {chromosom}\")\n",
    "\n",
    "#fig.supylabel(f'Variants per {round((reference_length/200)/1000)}kb.')\n",
    "\n",
    "alpha = 0.5\n",
    "\n",
    "for axs in (ax1, ax2, ax3, ax4):\n",
    "    __ = axs.twinx()\n",
    "    __.fill_between(list(range(sampling)),counts_editions, color= \"violet\",alpha= alpha)\n",
    "    __.set_ylabel(f'Editions per {round((reference_length/200)/1000,2)}kb.')\n",
    "    axs.set_xticks([])\n",
    "    axs.set_ylabel(f'Variants per {round((reference_length/200)/1000,2)}kb.')\n",
    "\n",
    "ax1.set_title(f\"Total variants for mgc graph (total: {sum(counts_A)})\")\n",
    "ax1.plot(counts_A_substitutions)\n",
    "ax1.plot([counts_A[i] - counts_A_substitutions[i] for i in range(sampling)])\n",
    "\n",
    "\n",
    "ax2.set_title(f\"Total variants for pggb graph (total: {sum(counts_B)})\")\n",
    "ax2.plot(counts_B_substitutions, label='substitutions')\n",
    "ax2.plot([counts_A[i]- counts_A_substitutions[i] for i in range(sampling)], label='other')\n",
    "\n",
    "ax2.legend(loc='upper right')\n",
    "\n",
    "# STABLE VARIANTS (shared variants across both graphs)\n",
    "\n",
    "ax3.set_title(f\"Shared substitution variants (total: {sum([counts_A_substitutions[i]- counts_A_private_substitution[i] for i in range(sampling)])})\")\n",
    "ax3.plot(shared_size_1, label='|v|$=$1')\n",
    "ax3.plot(shared_size_2_9, label='10$>$|v|$\\geq$2')\n",
    "ax3.plot(shared_size_10_49, label='50$>$|v|$\\geq$10')\n",
    "ax3.plot(shared_size_50, label='|v|$\\geq$50')\n",
    "\n",
    "ax3.legend(loc='upper right')\n",
    "\n",
    "ax4.set_title(f\"Shared other variants (total: {sum(shared_others)})\")\n",
    "ax4.plot(shared_others, label='others')\n",
    "\n",
    "plt.savefig(\"figures/variants_distribution_types.png\", bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
